src/NormalizingFlows.jl

```diff
@@ -9,20 +9,24 @@ using DiffResults
 import AbstractDifferentiation as AD
 
 include("train.jl")
+include("elbo.jl")
+
 function NF(
     vo,                                      # elbo, likelihood, f-div, STL, etc.. (how do we deal with this? it would require different input)
     flow::Bijectors.TransformedDistribution, # flow = T q₀, where T <: Bijectors.Bijector, q₀ reference dist that one can easily sample and compute logpdf
-    args...;                                # additional arguments for vo 
+    args...;                                 # additional arguments for vo 
     rng::AbstractRNG=Random.GLOBAL_RNG,
     max_iters::Int=1000,
     optimiser::Optimisers.AbstractRule=Optimisers.ADAM(),
     AD_backend::AD.AbstractBackend=AD.ZygoteBackend(),
 )
     # destruct flow for explicit access to the parameters
     # destructure can result in some overhead when the flow length is large
+    @info "desctructuring flow..."
     θ_flat, re = Flux.destructure(flow)
```

Comment:
This line can be really slow as mentioned in line 24; I tried the `flatten(x)` function written in https://github.com/TuringLang/AdvancedVI.jl/discussions/46#discussioncomment-5427443, which is also slow.

For reference, `Flux.destructure(flow)` with flow being 20-layer 2d planar layers takes ridiculously long time---way longer than the training time. I was wondering is there is a better way of handling the explicit parameterization in terms of performance.

Reply:
From my reply to that comment:

> > For reference, the excution of the following code took over an hour and I just kill the program for the lack of patience
> 
> This doesn't have anything to do with the usage of destructure; it's because you're doing `∘(...)`.
> 
> This then causes a `ComposedFunction` which is `n_layers` nested (`ComposedFunction(ComposedFunction(ComposedFunction(...)), ...)`; you get the gist). This blows up the compilation times.
> 
> For "large" compilations, you should instead use something like https://github.com/oschulz/FunctionChains.jl. Then you could use a `Vector` for to represent the composition which would make the compilation time waaaaay better (because now "iterate" over the recursion at runtime rather than at compile-time).
> 
> I also wonder if we could deal with this by subtyping `Function` due to https://docs.julialang.org/en/v1/devdocs/functions/#compiler-efficiency-issues thinking
> 
> But can discuss this issue more somewhere else +1